### PR TITLE
Type hints arelle.XmlUtil, and fix type annotations and full path from previous pr

### PR DESCRIPTION
**Pull request by selgamal**
#### Reason for change
Type hints

#### Description of change
Full path and annotations from pr #313
Type hints for XmlUtil.py

#### Following type ignores are added:
[L:725](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L725) type: ignore[arg-type] # ModelValue type hints
[L:728](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L728) type: ignore[no-untyped-call] # FunctionXs type hints
[L:730](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L730) type: ignore[no-untyped-call] # FunctionXs type hints
[L:734](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L734) type: ignore[no-untyped-call] # ModelDocument type hints
[L:857](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L857) type: ignore[return-value, operator] # ModelValue type hints
[L:867](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L867) type: ignore[return-value, operator]  # ModelValue type hints
[L:886](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L886) type: ignore[arg-type]  # recursive type see note line 883
[L:889](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L889) type: ignore[dict-item]  # recursive type note line 883
[L:948](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L948) type: ignore[name-defined]  # builtins
[L:985](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L985) type: ignore[no-untyped-call]  # ModelValue type hints
[L:1047](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L1047) type: ignore[union-attr] # ModelDocument type hints
[L:1128](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L1128) type: ignore[name-defined] # builtins
[L:1198](https://github.com/selgamal/Arelle/blob/07663239776dacfa45e3c690c1748c6042adec45/arelle/XmlUtil.py#L1198) type: ignore[name-defined] # builtins

#### Steps to Test

**review**:
@Arelle/arelle

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

**Comment by ggravlingen** [comment]
Great! I'm planning on resolving the review comments in the `ModelValue` PR over the weekend. It might help this PR to merge that first?

**Comment by selgamal** [comment]
> Great! I'm planning on resolving the review comments in the `ModelValue` PR over the weekend. It might help this PR to merge that first?

I may help, but there will always be another pass until all modules are typed and stable.

**Comment by ggravlingen** [comment]
> > Great! I'm planning on resolving the review comments in the `ModelValue` PR over the weekend. It might help this PR to merge that first?
> 
> 
> 
> It may help, but there will always be another pass until all modules are typed and stable. 

Makes sense.

If you are done with the edits on your end btw you should take the PR out of draft mode.

**Comment by selgamal** [comment]
> > > Great! I'm planning on resolving the review comments in the `ModelValue` PR over the weekend. It might help this PR to merge that first?
> > 
> > 
> > It may help, but there will always be another pass until all modules are typed and stable.
> 
> Makes sense.
> 
> If you are done with the edits on your end btw you should take the PR out of draft mode.

Yes done with the edits, I will take another look and convert it.

**Comment by selgamal** [comment]
This will need some updates after ModelObject merge, will help if we can re-run the tests on the current head

**Comment by ggravlingen** [comment]
> This will need some updates after ModelObject merge, will help if we can re-run the tests on the current head

Do you have previous experience with rebasing and resolving git conflicts? If not I'm happy to share what I normally do in situations like this (even though I'm no git expert).

**Comment by selgamal** [comment]
> > This will need some updates after ModelObject merge, will help if we can re-run the tests on the current head
> 
> Do you have previous experience with rebasing and resolving git conflicts? If not I'm happy to share what I normally do in situations like this (even though I'm no git expert).

Thanks a lot! I already tried that locally, I am waiting on reviews from the guys to incorporate all changes. When I get stuck I will you

File: arelle/XmlUtil.py
Comment (austinmatherne-wk) [inline arelle/XmlUtil.py:537]: I'm not sure about this change. The `PrototypeObject` base class doesn't have a qname property, but some (all?) subclasses of it do. It would be safer to keep the prototype isinstance check, but use `getattr` to avoid a potential attribute and mypy error.
```suggestion
            if isinstance(child, (ModelObject, PrototypeObject)) and ixTarget:
                childQname: QName | None = getattr(child, 'qname', None)
                if childQname:
                    _childNamespaceURI = childQname.namespaceURI
                    _childLocalName = childQname.localName
```

File: arelle/XmlUtil.py
Comment (austinmatherne-wk) [inline arelle/XmlUtil.py:1137]: I think this can be typed as `typing.TextIO`:
```suggestion
    writer: TextIO,
```

**Review by austinmatherne-wk — COMMENTED** [review]
A couple of the changes concern me, but I'm honestly impressed with how few issues there are. Really good job!

**Review by austinmatherne-wk — DISMISSED** [review]
There are merge conflicts with another PR that was just merged, but otherwise looks good.

**Review by austinmatherne-wk — APPROVED** [review]

**Pull request merged**
